### PR TITLE
Only show deprecation messages when new versions exist

### PR DESCRIPTION
**Pull request by komalali**
### Proposed changes

We now check if the suggested kind exists in the current k8s cluster version before showing a deprecation message.

It's kind of a pain to figure out which kinds were added in which spec version, so I wrote a little node script to generate a list for me. Not sure if it's useful to put it somewhere, but I've pasted it here and open to ideas.

```javascript
const fetch = require("node-fetch");

// Fetches the spec for the specified version and returns an array of keys for all the 
// definitions with the 'x-kubernetes-group-version-kind' property (signifying a GVK).
async function getSpec(version) {
    const response = await fetch(`https://raw.githubusercontent.com/kubernetes/kubernetes/release-1.${version}/api/openapi-spec/swagger.json`);
    const json = await response.json();
    const types = [];
    Object.entries(json.definitions).forEach(([key, val]) => {
        if (val['x-kubernetes-group-version-kind']) {
            types.push(key)
        }
    })
    return types;
}

// Map through the versions and collect all the resources that are not present in the 
// previous version.
(async function main() {
    const versions = [13, 14, 15, 16, 17, 18];
    const versionMap = {};
    const results = await Promise.all(versions.map(getSpec));
    results.forEach((result, index) => versionMap[`v${versions[index]}`] = result);

    const newResources = {};

    function getNewResources(version) {
        const newInVersion = [];
        versionMap[`v${version}`].forEach(resource => {
            if (!versionMap[`v${version - 1}`].includes(resource)) {
                newInVersion.push(resource);
            }
        })
        newResources[`v${version}`] = newInVersion;
    }

    for (let i=1; i<versions.length; i++) {
        getNewResources(versions[i])
    }

    console.log(newResources);
})()
```

### Related issues (optional)

Fixes: #1168

### CODE REVIEW

File: provider/pkg/kinds/deprecated.go [hunk 1]
```diff
@@ -66,48 +67,146 @@ import (
...
+	case StorageV1:
+		switch k {
+		case CSINode, CSINodeList:
+			return &v117
+		case CSIDriver, CSIDriverList:
+			return &v118
+		}
+	}
+
+	return &cluster.ServerVersion{Major: 1, Minor: 0}
```
Comment (lblackstone) [inline provider/pkg/kinds/deprecated.go:172]: It looks like this is only checking back to v1.14, so this is not quite correct. Can you extend this logic back to v1.10? That would cover the last two years of releases, which should account for the majority of cases.

File: provider/cmd/pulumi-gen-kubernetes/main.go [hunk 2]
```diff
@@ -319,9 +319,6 @@ func genK8sResourceTypes(pkg *schema.Package) {
 		contract.Assert(len(parts) == 3)
 
 		groupVersion, kind := parts[1], parts[2]
-		if strings.HasSuffix(kind, "List") {
```
Comment (lblackstone) [inline provider/cmd/pulumi-gen-kubernetes/main.go:322]: I realized that we're not handling List deprecations for a few cases. Could you update that logic to include the corresponding List types as well? https://github.com/pulumi/pulumi-kubernetes/pull/1182/files#diff-239b11d00160719500c6fea14202bd9eL150-L157

**Comment by lblackstone** [comment]
> It's kind of a pain to figure out which kinds were added in which spec version, so I wrote a little node script to generate a list for me. Not sure if it's useful to put it somewhere, but I've pasted it here and open to ideas.

We may eventually want to turn this into something that runs as part of code generation (translated to Go). Can you open an issue to track that?

File: provider/pkg/kinds/deprecated.go (hunk 1 above)
Comment (komalali) [reply provider/pkg/kinds/deprecated.go:172]: Sounds good will do.

File: provider/pkg/kinds/deprecated.go [hunk 3]
```diff
@@ -66,48 +67,194 @@ import (
...
+	}
+
+	// We extends this logic back to v1.10, so for all other kinds we return 1.9, meaning that anyone
+	// on a 1.9 or earlier cluster will not see deprecation messages.
+	return &v19
+}
+
+// ExistsInVersion returns true if the given GVK exists in the given k8s version.
+func ExistsInVersion(gvk *schema.GroupVersionKind, version *cluster.ServerVersion) bool {
+	if gvk.Empty() {
```
Comment (lblackstone) [inline provider/pkg/kinds/deprecated.go:222]: Need to check for nil inputs as well to avoid a possible panic.

**Pull request merged**